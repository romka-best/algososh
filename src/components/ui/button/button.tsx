import React from 'react';
import cn from 'classnames';

import loaderIcon from '../../../images/icons/loader.svg';
import { Direction } from '../../../types/direction';

import { AscendingIcon } from '../icons/ascending-icon';
import { DescendingIcon } from '../icons/descending-icon';

import styles from './button.module.css';

interface ButtonProps extends React.HTMLProps<HTMLButtonElement> {
    text?: string;
    type?: 'button' | 'submit' | 'reset';
    sorting?: Direction;
    linkedList?: 'small' | 'big';
    isLoader?: boolean;
    extraClass?: string;
}

export const Button: React.FC<ButtonProps> = (
    {
        text,
        extraClass = '',
        type = 'button',
        isLoader = false,
        sorting,
        linkedList,
        disabled,
        ...rest
    }
) => {
    const currentIcon =
        sorting === Direction.Ascending ? <AscendingIcon/> : <DescendingIcon/>;
    const className = cn('text',
        'text_type_button',
        'text_color_primary',
        styles.button,
        linkedList && styles[linkedList],
        isLoader && styles.loader,
        extraClass,
    );

    return (
        <button
            className={ className }
            type={ type }
            disabled={ isLoader || disabled }
            { ...rest }
        >
            { isLoader ? (
                <img className={ styles.loader_icon } src={ loaderIcon } alt="Загрузка."/>
            ) : (
                <>
                    { sorting && currentIcon }
                    <p className={ cn('text', sorting && 'ml-5') }>{ text }</p>
                </>
            ) }
        </button>
    );
};
