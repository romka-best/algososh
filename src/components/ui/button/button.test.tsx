import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';

import { Direction } from '../../../types/direction';

import { Button } from './button';

describe('Button', () => {
    describe('Отображение', () => {
        it('С текстом', () => {
            const button = renderer
                .create(
                    <Button
                        text="Жмяк!"
                    />
                )
                .toJSON();

            expect(button).toMatchSnapshot();
        });

        it('Без текста', () => {
            const button = renderer
                .create(
                    <Button/>
                )
                .toJSON();

            expect(button).toMatchSnapshot();
        });

        it('Заблокированная', () => {
            const button = renderer
                .create(
                    <Button
                        disabled={ true }
                    />
                )
                .toJSON();

            expect(button).toMatchSnapshot();
        });

        it('С индикацией загрузки', () => {
            const button = renderer
                .create(
                    <Button
                        isLoader={ true }
                    />
                )
                .toJSON();

            expect(button).toMatchSnapshot();
        });

        describe('Разные типы', () => {
            it('По умолчанию', () => {
                const button = renderer
                    .create(
                        <Button/>
                    )
                    .toJSON();

                expect(button).toMatchSnapshot();
            });

            it('Submit', () => {
                const button = renderer
                    .create(
                        <Button
                            type="submit"
                        />
                    )
                    .toJSON();

                expect(button).toMatchSnapshot();
            });

            it('Reset', () => {
                const button = renderer
                    .create(
                        <Button
                            type="reset"
                        />
                    )
                    .toJSON();

                expect(button).toMatchSnapshot();
            });
        });

        describe('Разная сортировка', () => {
            it('По возрастанию', () => {
                const button = renderer
                    .create(
                        <Button
                            sorting={ Direction.Ascending }
                        />
                    )
                    .toJSON();

                expect(button).toMatchSnapshot();
            });

            it('По убыванию', () => {
                const button = renderer
                    .create(
                        <Button
                            sorting={ Direction.Descending }
                        />
                    )
                    .toJSON();

                expect(button).toMatchSnapshot();
            });
        });

        describe('Разный размер', () => {
            it('Маленький', () => {
                const button = renderer
                    .create(
                        <Button
                            linkedList="small"
                        />
                    )
                    .toJSON();

                expect(button).toMatchSnapshot();
            });

            it('Большой', () => {
                const button = renderer
                    .create(
                        <Button
                            linkedList="big"
                        />
                    )
                    .toJSON();

                expect(button).toMatchSnapshot();
            });
        });

        it('С дополнительным классом', () => {
            const button = renderer
                .create(
                    <Button
                        extraClass="newClass"
                    />
                )
                .toJSON();

            expect(button).toMatchSnapshot();
        });
    });

    describe('Логика', () => {
        it('Вызов колбека при клике', () => {
            console.log = jest.fn();
            const onClick = () => {
                console.log('Жмяк!');
            };

            render(
                <Button
                    title="Жмяк!"
                    onClick={ onClick }
                />
            );

            const button = screen.getByTitle('Жмяк!');

            fireEvent.click(button);
            expect(console.log).toHaveBeenCalledWith('Жмяк!');
        });
    });
});
