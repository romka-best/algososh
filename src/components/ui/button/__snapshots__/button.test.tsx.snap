// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Отображение Без текста 1`] = `
<button
  className="text text_type_button text_color_primary button"
  type="button"
>
  <p
    className="text"
  />
</button>
`;

exports[`Button Отображение Заблокированная 1`] = `
<button
  className="text text_type_button text_color_primary button"
  disabled={true}
  type="button"
>
  <p
    className="text"
  />
</button>
`;

exports[`Button Отображение Разная сортировка По возрастанию 1`] = `
<button
  className="text text_type_button text_color_primary button"
  type="button"
>
  <svg
    fill="inherit"
    height="12"
    viewBox="0 0 20 12"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M1 0C0.447715 0 0 0.447715 0 1C0 1.55228 0.447715 2 1 2H11C11.5523 2 12 1.55228 12 1C12 0.447715 11.5523 0 11 0H1ZM0 6C0 5.44772 0.447715 5 1 5H15C15.5523 5 16 5.44772 16 6C16 6.55228 15.5523 7 15 7H1C0.447715 7 0 6.55228 0 6ZM0 11C0 10.4477 0.447715 10 1 10H19C19.5523 10 20 10.4477 20 11C20 11.5523 19.5523 12 19 12H1C0.447716 12 0 11.5523 0 11Z"
      fill="inherit"
      fillOpacity="0.85"
    />
  </svg>
  <p
    className="text ml-5"
  />
</button>
`;

exports[`Button Отображение Разная сортировка По убыванию 1`] = `
<button
  className="text text_type_button text_color_primary button"
  type="button"
>
  <svg
    fill="inherit"
    height="12"
    viewBox="0 0 20 12"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M0 1C0 0.44772 0.44772 0 1 0H19C19.5523 0 20 0.44772 20 1C20 1.55228 19.5523 2 19 2H1C0.44772 2 0 1.55228 0 1ZM0 6C0 5.4477 0.44772 5 1 5H15C15.5523 5 16 5.4477 16 6C16 6.5523 15.5523 7 15 7H1C0.44772 7 0 6.5523 0 6ZM1 10C0.44772 10 0 10.4477 0 11C0 11.5523 0.44772 12 1 12H11C11.5523 12 12 11.5523 12 11C12 10.4477 11.5523 10 11 10H1Z"
      fill="inherit"
      fillOpacity="0.85"
      fillRule="evenodd"
    />
  </svg>
  <p
    className="text ml-5"
  />
</button>
`;

exports[`Button Отображение Разные типы Reset 1`] = `
<button
  className="text text_type_button text_color_primary button"
  type="reset"
>
  <p
    className="text"
  />
</button>
`;

exports[`Button Отображение Разные типы Submit 1`] = `
<button
  className="text text_type_button text_color_primary button"
  type="submit"
>
  <p
    className="text"
  />
</button>
`;

exports[`Button Отображение Разные типы По умолчанию 1`] = `
<button
  className="text text_type_button text_color_primary button"
  type="button"
>
  <p
    className="text"
  />
</button>
`;

exports[`Button Отображение Разный размер Большой 1`] = `
<button
  className="text text_type_button text_color_primary button big"
  type="button"
>
  <p
    className="text"
  />
</button>
`;

exports[`Button Отображение Разный размер Маленький 1`] = `
<button
  className="text text_type_button text_color_primary button small"
  type="button"
>
  <p
    className="text"
  />
</button>
`;

exports[`Button Отображение С дополнительным классом 1`] = `
<button
  className="text text_type_button text_color_primary button newClass"
  type="button"
>
  <p
    className="text"
  />
</button>
`;

exports[`Button Отображение С индикацией загрузки 1`] = `
<button
  className="text text_type_button text_color_primary button loader"
  disabled={true}
  type="button"
>
  <img
    alt="Загрузка."
    className="loader_icon"
    src="loader.svg"
  />
</button>
`;

exports[`Button Отображение С текстом 1`] = `
<button
  className="text text_type_button text_color_primary button"
  type="button"
>
  <p
    className="text"
  >
    Жмяк!
  </p>
</button>
`;
